staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,256 @@
+	// resources (channel and goroutine). The caller should also close q to
+	// stop the background goroutine.
+	cancel context.CancelFunc
+}
+
+// manager implements the Manager interface.
+type manager struct {
+	// updateQueues holds a CRD UID to updateQueue map. Each CRD has its
+	// own queue of StorageVersion updates, and a goroutine which processes
+	// the pending updates in the queue. The manager sends update requests
+	// to the queue.
+	updateQueues map[types.UID]*updateQueue
+	// client is the client interface that manager uses to update
+	// StorageVersion objects.
+	client genericstorageversion.Client
+	// apiserverID is the ID of the apiserver that invokes this manager.
+	apiserverID string
+}
+
+// NewManager creates a CRD StorageVersion Manager.
+func NewManager(client genericstorageversion.Client, apiserverID string) Manager {
+	return &manager{
+		updateQueues: make(map[types.UID]*updateQueue),
+		client:       client,
+		apiserverID:  apiserverID,
+	}
+}
+
+// UpdateStorageVersionFor queues a StorageVesrion update for the given
+// CRD and returns immediately. Optionally, the caller may specify a
+// non-nil waitCh and/or a non-nil processedCh and a non-nil abortedCh.
+// A non-nil waitCh will block the StorageVersion update until waitCh is
+// closed.
+// The manager will close the non-nil processedCh if it finished
+// processing the StorageVersion update (note that the update can either
+// succeeded or failed), or close the non-nil abortedCh if it aborted
+// processing the update.
+func (m *manager) UpdateStorageVersionFor(crd *apiextensionsv1.CustomResourceDefinition,
+	waitCh <-chan struct{}, processedCh, abortedCh chan<- struct{}) {
+	q := m.getOrCreateUpdateQueue(crd.UID)
+	// When the channel buffer is full, writing to the channel becomes
+	// blocking. Here we give up updating storage version for this CRD and
+	// print a log, so that we can return immediately and not block the
+	// caller.
+	if len(q) == updateQueueBufferSize {
+		// TODO(roycaihw): use warning instead of info when StorageVersionAPI
+		// graduates to beta/GA
+		klog.V(2).Infof("Skipping the storage version update for CRD with UID %v due to the queue being full (queue size: %v).",
+			crd.UID, updateQueueBufferSize)
+		if processedCh != nil {
+			close(processedCh)
+		}
+		return
+	}
+	q <- &update{
+		crd:         crd,
+		waitCh:      waitCh,
+		processedCh: processedCh,
+		abortedCh:   abortedCh,
+	}
+}
+
+// getOrCreateUpdateQueue returns the channel for the given UID, or create a new
+// one and a new goroutine if necessary. The goroutine keeps processing updates
+// until the channel is closed.
+func (m *manager) getOrCreateUpdateQueue(uid types.UID) chan<- *update {
+	if queue, ok := m.updateQueues[uid]; ok {
+		return queue.q
+	}
+
+	queue := make(chan *update, updateQueueBufferSize)
+	ctx, cancel := context.WithCancel(context.TODO())
+	m.updateQueues[uid] = &updateQueue{
+		q:      queue,
+		cancel: cancel,
+	}
+	go func() {
```

Comment:
this need panic handling. Otherwise updates for a UID can be blocker forever (until process restart).